### PR TITLE
[sc-22364] Create GitHub template

### PR DESCRIPTION
**Pull request by arooxa**
This template supports all MVP entities that are listed in this story [sc-15039](https://app.shortcut.com/sgnl/story/15039/github-as-a-system-of-record)
For Reviewers - this was tested successfully in dev

[sc-22364](https://app.shortcut.com/sgnl/story/22364/create-github-catalog-template)

### CODE REVIEW

File: github.sgnl.yaml [hunk 1]
```diff
@@ -0,0 +1,1118 @@
...
+    apiCallMinInterval: 1
+    pageSize: 100
+    pagesOrderedById: false
+    attributes:
+      - name: email
+        externalId: email
+        type: String
+        indexed: true
+        uniqueId: true
+relationships:
```
Comment (isabella-sgnl) [inline github.sgnl.yaml:1058]: @arooxa Can we simplify these relationship names? Some general ideas included below:
- "Parent of Repository Collaborator" to "RepositoryCollaborator in the display name
- "RepositoryToOrganization" can be simplified to "OrganizationRepository" 
- "TeamToOrganization" to "OrganizationTeam"
- "UserToOrganizationUser" to "OrganizationUser"


Aim for this is to be consistent with other relationship names, Jira and Salesforce pictured below:
Jira
Screenshot 2024-03-11 at 5 52 03 PM
SFDC
Screenshot 2024-03-11 at 5 52 23 PM

File: github.sgnl.yaml [hunk 2]
```diff
@@ -0,0 +1,1099 @@
...
+        externalId: viewerIsFollowing
+        type: Bool
+      - name: websiteUrl
+        externalId: websiteUrl
+        type: String
+relationships:
+  RepositoryToOrganization:
+    displayName: OrganizationRepository
+    name: MemberOf
+    fromAttribute: Repository-orgId
```
Comment (isabella-sgnl) [inline github.sgnl.yaml:1044]: Is this formatting standard across different SoRs in our adapters? Wondering if the camelCase/naming conventions here should make this more like "repository-orgId" or just "repositoryOrgId." No strong stance here on the formatting, but just wanted to double check that this is consistent across this template and other SoRs.

File: github.sgnl.yaml [hunk 3]
```diff
@@ -0,0 +1,1099 @@
...
+        type: String
+        indexed: true
+        uniqueId: true
+        relationshipAttributeId: TeamMember-id
+      # Note: Role represents the TeamMember's role on the team.
+      # https://docs.github.com/en/graphql/reference/enums#teammemberrole
+      - name: role
+        externalId: role
+        type: String
+      - name: avatarUrl
```
Comment (r-rakshith) [inline github.sgnl.yaml:431]: Apart from `id/login/email` (a unique identifier for a User) and `role` (in the context of a Team), if there's no attributes unique to a team-member as compared to a user - then we can exclude these attributes altogether.

Team member would be same-as User right?
User would be a superset of Team member right?

File: github.sgnl.yaml [hunk 4]
```diff
@@ -0,0 +1,1099 @@
...
+  # TeamRepository is a Repository entity type: https://docs.github.com/en/graphql/reference/objects#repository
+  TeamRepository:
+    parent: Team
+    displayName: TeamRepository
+    externalId: $.repositories.edges
+    description: Entity representing a GitHub repository assigned to a team.
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
Comment (r-rakshith) [inline github.sgnl.yaml:534]:
```suggestion
```

File: github.sgnl.yaml [hunk 5]
```diff
@@ -0,0 +1,1099 @@
...
+    attributes:
+      - name: id
+        externalId: $.node.id
+        type: String
+        indexed: true
+        uniqueId: true
+        relationshipAttributeId: TeamRepository-id
+        # Permission is the Team's permission level for this repository
+        # https://docs.github.com/en/graphql/reference/enums#repositorypermission
+      - name: permission
```
Comment (r-rakshith) [inline github.sgnl.yaml:546]: Same comment as above. Ingest only those attributes that `Repository` entity does not have but this one has in the context of a Team (along with unique identifier ofcourse). 

TeamRepository would be same-as Repository right?
Repository would be a superset of TeamRepository right?

File: github.sgnl.yaml [hunk 6]
```diff
@@ -0,0 +1,1099 @@
...
+        externalId: $.node.viewerPossibleCommitEmails
+        type: String
+      - name: webCommitSignoffRequired
+        externalId: $.node.webCommitSignoffRequired
+        type: Bool
+  # https://docs.github.com/en/graphql/reference/objects#repository
+  Repository:
+    displayName: Repository
+    externalId: Repository
+    description: Repository entity in GitHub
```
Comment (r-rakshith) [inline github.sgnl.yaml:718]: Update description based on the format decided above.

File: github.sgnl.yaml [hunk 7]
```diff
@@ -0,0 +1,1099 @@
...
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
Comment (r-rakshith) [inline github.sgnl.yaml:722]:
```suggestion
```

File: github.sgnl.yaml [hunk 8]
```diff
@@ -0,0 +1,1099 @@
...
+        type: Bool
+  RepositoryCollaborator:
+    parent: Repository
+    displayName: RepositoryCollaborator
+    externalId: $.collaborators.edges
+    description: Entity representing the relationship between a Repository and Collaborator entity in GitHub
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
Comment (r-rakshith) [inline github.sgnl.yaml:916]:
```suggestion
```

File: github.sgnl.yaml (hunk 2 above)
Comment (r-rakshith) [reply github.sgnl.yaml:1044]: Let's decide on one of `lowerCamelCase` or `UpperCamelCase` here. `-`'s may cause issues down the line.

**Review by r-rakshith — CHANGES_REQUESTED** [review]
Some changes requested throughout the file.

- Remove all sync* and apiCall* settings on entity. Let it inherit this from datasource defaults.
- I don't think we should ingest *HTML attributes. Same goes for *Url attributes too. @isabella-sgnl to confirm if there's any usecases for these attributes.
- Deduplicate attributes across Entities of the same type -  TeamMember<>User for example as mentioned in the comments.
- Descriptions to follow a format if possible
- Styling attribute aliases.

File: github.sgnl.yaml (hunk 2 above)
Comment (arooxa) [reply github.sgnl.yaml:1044]: will change to upper camel case

File: github.sgnl.yaml [hunk 9]
```diff
@@ -0,0 +1,1099 @@
...
+entities:
+  # https://docs.github.com/en/graphql/reference/objects#organization
+  Organization:
+    displayName: Organization
+    externalId: Organization
+    description: Organization entity in GitHub.
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:26]: fixed

File: github.sgnl.yaml [hunk 10]
```diff
@@ -0,0 +1,1099 @@
...
...Wxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPgo8cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGNsaXAtcnVsZT0iZXZlbm9kZCIgZD0iTTggMEMzLjU4IDAgMCAzLjU4IDAgOEMwIDExLjU0IDIuMjkgMTQuNTMgNS40NyAxNS41OUM1Ljg3IDE1LjY2IDYuMDIgMTUuNDIgNi4wMiAxNS4yMUM2LjAyIDE1LjAyIDYuMDEgMTQuMzkgNi4wMSAxMy43MkM0IDE0LjA5IDMuNDggMTMuMjMgMy4zMiAxMi43OEMzLjIzIDEyLjU1IDIuODQgMTEuODQgMi41IDExLjY1QzIuMjIgMTEuNSAxLjgyIDExLjEzIDIuNDkgMTEuMTJDMy4xMiAxMS4xMSAzLjU3IDExLjcgMy43MiAxMS45NEM0LjQ0IDEzLjE1IDUuNTkgMTIuODEgNi4wNSAxMi42QzYuMTIgMTIuMDggNi4zMyAxMS43MyA2LjU2IDExLjUzQzQuNzggMTEuMzMgMi45MiAxMC42NCAyLjkyIDcuNThDMi45MiA2LjcxIDMuMjMgNS45OSAzLjc0IDUuNDNDMy42NiA1LjIzIDMuMzggNC40MSAzLjgyIDMuMzFDMy44MiAzLjMxIDQuNDkgMy4xIDYuMDIgNC4xM0M2LjY2IDMuOTUgNy4zNCAzLjg2IDguMDIgMy44NkM4LjcgMy44NiA5LjM4IDMuOTUgMTAuMDIgNC4xM0MxMS41NSAzLjA5IDEyLjIyIDMuMzEgMTIuMjIgMy4zMUMxMi42NiA0LjQxIDEyLjM4IDUuMjMgMTIuMyA1LjQzQzEyLjgxIDUuOTkgMTMuMTIgNi43IDEzLjEyIDcuNThDMTMuMTIgMTAuNjUgMTEuMjUgMTEuMzMgOS40NyAxMS41M0M5Ljc2IDExLjc4IDEwLjAxIDEyLjI2IDEwLjAxIDEzLjAxQzEwLjAxIDE0LjA4IDEwIDE0Ljk0IDEwIDE1LjIxQzEwIDE1LjQyIDEwLjE1IDE1LjY3IDEwLjU1IDE1LjU5QzEzLjcxIDE0LjUzIDE2IDExLjUzIDE2IDhDMTYgMy41OCAxMi40MiAwIDggMFoiIHRyYW5zZm9ybT0ic2NhbGUoNjQpIiBmaWxsPSIjMUIxRjIzIi8+Cjwvc3ZnPgo="
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:3]: fixed

File: github.sgnl.yaml [hunk 11]
```diff
@@ -0,0 +1,1099 @@
...
+      - name: viewerIsAMember
+        externalId: viewerIsAMember
+        type: Bool
+      - name: websiteUrl
+        externalId: websiteUrl
+        type: String
+  OrganizationUser:
+    displayName: OrganizationUser
+    externalId: OrganizationUser
+    description: Entity representing the relationship between an Organization and User entity in GitHub
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:151]: fixed

File: github.sgnl.yaml [hunk 12]
```diff
@@ -0,0 +1,1099 @@
...
+        type: DateTime
+      - name: databaseId
+        externalId: databaseId
+        type: Int64
+      - name: description
+        externalId: description
+        type: String
+      - name: descriptionHTML
+        externalId: descriptionHTML
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:69]: fixed

File: github.sgnl.yaml [hunk 13]
```diff
@@ -0,0 +1,1099 @@
...
+        externalId: websiteUrl
+        type: String
+  OrganizationUser:
+    displayName: OrganizationUser
+    externalId: OrganizationUser
+    description: Entity representing the relationship between an Organization and User entity in GitHub
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY 
+    apiCallMinInterval: 1
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:155]: fixed

File: github.sgnl.yaml [hunk 14]
```diff
@@ -0,0 +1,1099 @@
...
+        type: String
+  # https://docs.github.com/en/graphql/reference/objects#user
+  User:
+    displayName: User
+    externalId: User
+    description: User entity in GitHub.
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:185]: fixed

File: github.sgnl.yaml [hunk 15]
```diff
@@ -0,0 +1,1099 @@
...
+    attributes:
+      - name: id
+        externalId: id
+        type: String
+        indexed: true
+        uniqueId: true
+        relationshipAttributeId: User-id
+      - name: avatarUrl
+        externalId: avatarUrl
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:197]: fixed

File: github.sgnl.yaml [hunk 16]
```diff
@@ -0,0 +1,1099 @@
...
+      - name: bioHTML
+        externalId: bioHTML
+        type: String
+      - name: company
+        externalId: company
+        type: String
+      - name: companyHTML
+        externalId: companyHTML
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:209]: fixed

File: github.sgnl.yaml [hunk 17]
```diff
@@ -0,0 +1,1099 @@
...
+        relationshipAttributeId: User-id
+      - name: avatarUrl
+        externalId: avatarUrl
+        type: String
+      - name: bio
+        externalId: bio
+        type: String
+      - name: bioHTML
+        externalId: bioHTML
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:203]: fixed

File: github.sgnl.yaml [hunk 18]
```diff
@@ -0,0 +1,1099 @@
...
+        type: Bool
+      - name: viewerIsFollowing
+        externalId: viewerIsFollowing
+        type: Bool
+      - name: websiteUrl
+        externalId: websiteUrl
+        type: String
+  UserOVDE:
+    parent: OrganizationUser
+    displayName: UserOVDE
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:290]: fixed

File: github.sgnl.yaml [hunk 19]
```diff
@@ -0,0 +1,1099 @@
...
+        uniqueId: true 
+  # https://docs.github.com/en/graphql/reference/objects#team  
+  Team:
+    displayName: Team
+    externalId: Team
+    description: Team entity in GitHub.
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:313]: fixed

File: github.sgnl.yaml [hunk 20]
```diff
@@ -0,0 +1,1099 @@
...
+        type: String
+      # Note: This attribute has been added in post-processing and is not part of the Team entity.
+      - name: orgId
+        externalId: orgId
+        type: String
+        indexed: true
+        relationshipAttributeId: Team-orgId
+      - name: avatarUrl
+        externalId: avatarUrl
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:334]: fixed

File: github.sgnl.yaml [hunk 21]
```diff
@@ -0,0 +1,1099 @@
...
+  # TeamMember is a User entity type: https://docs.github.com/en/graphql/reference/objects#user
+  TeamMember:
+    parent: Team
+    displayName: TeamMember
+    externalId: $.members.edges
+    description: Entity representing a GitHub team each user belongs to.
+    syncFrequency: HOURLY
+    syncMinInterval: 1
+    apiCallFrequency: SECONDLY
+    apiCallMinInterval: 1
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:416]: fixed

File: github.sgnl.yaml [hunk 22]
```diff
@@ -0,0 +1,1099 @@
...
+      - name: viewerCanAdminister 
+        externalId: viewerCanAdminister
+        type: Bool
+      - name: viewerCanSubscribe
+        externalId: viewerCanSubscribe
+        type: Bool
+  # Fetched from the 'members' attribute belonging to the Teams entity.
+  # TeamMember entity is for creating relationships between Users and Teams and does not support organizationVerifiedDomainEmails.
+  # To retrieve organizationVerifiedDomainEmails, declare it as an attribute for the User entity.
+  # TeamMember is a User entity type: https://docs.github.com/en/graphql/reference/objects#user
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:407]: fixed

File: github.sgnl.yaml (hunk 3 above)
Comment (arooxa) [reply github.sgnl.yaml:431]: fixed

File: github.sgnl.yaml (hunk 4 above)
Comment (arooxa) [reply github.sgnl.yaml:534]: fixed

File: github.sgnl.yaml [hunk 23]
```diff
@@ -0,0 +1,1099 @@
...
+        indexed: true
+        relationshipAttributeId: OrganizationUser-userId
+      - name: role
+        externalId: role
+        type: String
+  # https://docs.github.com/en/graphql/reference/objects#user
+  User:
+    displayName: User
+    externalId: User
+    description: User entity in GitHub.
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:181]: fixed

File: github.sgnl.yaml [hunk 24]
```diff
@@ -0,0 +1,1099 @@
...
+      # Note: This attribute has been added in post-processing and is not part of the Team entity.
+      - name: enterpriseId
+        externalId: enterpriseId
+        type: String
+      # Note: This attribute has been added in post-processing and is not part of the Team entity.
+      - name: orgId
+        externalId: orgId
+        type: String
+        indexed: true
+        relationshipAttributeId: Team-orgId
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:331]: fixed

File: github.sgnl.yaml (hunk 5 above)
Comment (arooxa) [reply github.sgnl.yaml:546]: fixed

File: github.sgnl.yaml (hunk 6 above)
Comment (arooxa) [reply github.sgnl.yaml:718]: fixed

File: github.sgnl.yaml (hunk 7 above)
Comment (arooxa) [reply github.sgnl.yaml:722]: fixed

File: github.sgnl.yaml (hunk 8 above)
Comment (arooxa) [reply github.sgnl.yaml:916]: fixed

File: github.sgnl.yaml [hunk 25]
```diff
@@ -0,0 +1,1099 @@
...
+# Enterprise Server: {hostname}/api/graphql
+address: "https://github-enterprise-server.sgnldemos.com/api/graphql"
+defaultSyncFrequency: HOURLY
+defaultSyncMinInterval: 1
+defaultApiCallFrequency: SECONDLY
+defaultApiCallMinInterval: 1
+type: "GitHub-1.0.0"
+adapterConfig: "ewogICAgImVudGVycHJpc2VTbHVnIjogIlNHTkwiCn0=" # { enterpriseSlug: "SGNL" }
+auth:
+  - bearer: # https://docs.github.com/en/enterprise-server@3.10/authentication/keeping-your-account-and-data-secure/about-authentication-to-github
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:15]: fixed

File: github.sgnl.yaml [hunk 26]
```diff
@@ -0,0 +1,1099 @@
...
+description: "GitHub as a System of Record"
+# Enterprise Cloud: https://api.github.com/graphql 
+# Enterprise Server: {hostname}/api/graphql
+address: "https://github-enterprise-server.sgnldemos.com/api/graphql"
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml]: fixed

File: github.sgnl.yaml [hunk 27]
```diff
@@ -0,0 +1,1099 @@
...
+      - name: id
+        externalId: id
+        type: String
+        indexed: true
+        uniqueId: true
+        relationshipAttributeId: Organization-id
+      # Note: This attribute has been added in post-processing and is not part of the Organization entity.
+      - name: enterpriseId
+        externalId: enterpriseId
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml]: fixed

File: github.sgnl.yaml [hunk 28]
```diff
@@ -0,0 +1,1099 @@
...
+        indexed: true
+        relationshipAttributeId: OrganizationUser-orgId
+      - name: userId
+        externalId: $.node.id
+        type: String
+        indexed: true
+        relationshipAttributeId: OrganizationUser-userId
+      - name: role
+        externalId: role
+        type: String
```
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:147]: I think we can leave this for now, and add it later if there's a use-case. WDYT?

[the remark this answers is not in the document]
Comment (isabella-sgnl) [reply github.sgnl.yaml:147]: I can see user role being useful (@arooxa our snippets for other SoRs include simple ones like User is on DevOps team or has role of Customer Support), agreed on indexing if that sounds good.

**Comment by isabella-sgnl** [comment]
> Some changes requested throughout the file.
> 
> * Remove all sync* and apiCall* settings on entity. Let it inherit this from datasource defaults.
> * I don't think we should ingest *HTML attributes. Same goes for *Url attributes too. @isabella-sgnl to confirm if there's any usecases for these attributes.
> * Deduplicate attributes across Entities of the same type -  TeamMember<>User for example as mentioned in the comments.
> * Descriptions to follow a format if possible
> * Styling attribute aliases.

Removing HTML and URL attributes SGTM, don't forsee a use-case for those.

File: github.sgnl.yaml (hunk 28 above)
[the remark this answers is not in the document]
Comment (arooxa) [reply github.sgnl.yaml:147]: Ok, sounds good. In that case, I should also add indexing to the role and permission attributes in RepositoryCollaborator, TeamMember, and TeamRepository right?

[the remark this answers is not in the document]
Comment (isabella-sgnl) [reply github.sgnl.yaml:147]: Yes, let's add indexing for those attributes.

**Comment by marc-sgnl** [comment]
adapterConfig seems inconsistent here:

> adapterConfig: "ewogICAgImVudGVycHJpc2VTbHVnIjogIlNHTkwiCn0=" # { enterpriseSlug: "SGNL" }

the `enterpriseSlug` is customer-enterprise specific and will need to be entered. We treat other placeholders of this type in this and other templates with a curly brace treatment, suggest that this should be more akin to other placeholders, e.g. `{ enterpriseSlug: "{{Your Enterprise Name}}" }`

**Comment by arooxa** [comment]
> adapterConfig seems inconsistent here:
> 
> > adapterConfig: "ewogICAgImVudGVycHJpc2VTbHVnIjogIlNHTkwiCn0=" # { enterpriseSlug: "SGNL" }
> 
> the `enterpriseSlug` is customer-enterprise specific and will need to be entered. We treat other placeholders of this type in this and other templates with a curly brace treatment, suggest that this should be more akin to other placeholders, e.g. `{ enterpriseSlug: "{{Your Enterprise Name}}" }`

Good catch, just updated this.

**Review by r-rakshith — APPROVED** [review]
LGTM

**Pull request merged**